cpp/src/arrow/compute/kernels/aggregate_basic_internal.h

```diff
@@ -0,0 +1,303 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/compute/kernels/aggregate_internal.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/util/align_util.h"
+#include "arrow/util/bit_block_counter.h"
+
+namespace arrow {
+namespace compute {
+namespace aggregate {
+
+struct ScalarAggregator : public KernelState {
+  virtual void Consume(KernelContext* ctx, const ExecBatch& batch) = 0;
+  virtual void MergeFrom(KernelContext* ctx, const KernelState& src) = 0;
+  virtual void Finalize(KernelContext* ctx, Datum* out) = 0;
+};
+
+void AddBasicAggKernels(KernelInit init,
+                        const std::vector<std::shared_ptr<DataType>>& types,
+                        std::shared_ptr<DataType> out_ty, ScalarAggregateFunction* func);
+
+// ----------------------------------------------------------------------
+// Sum implementation
+
+template <int64_t kRoundSize, typename ArrowType>
+struct SumState {
+  using SumType = typename FindAccumulatorType<ArrowType>::Type;
+  using ThisType = SumState<kRoundSize, ArrowType>;
+  using T = typename TypeTraits<ArrowType>::CType;
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+
+  ThisType operator+(const ThisType& rhs) const {
+    return ThisType(this->count + rhs.count, this->sum + rhs.sum);
+  }
+
+  ThisType& operator+=(const ThisType& rhs) {
+    this->count += rhs.count;
+    this->sum += rhs.sum;
+
+    return *this;
+  }
+
+ public:
+  void Consume(const Array& input) {
+    const ArrayType& array = static_cast<const ArrayType&>(input);
+    if (input.null_count() == 0) {
+      (*this) += ConsumeDense(array);
+    } else {
+      (*this) += ConsumeSparse(array);
+    }
+  }
+
+  size_t count = 0;
+  typename SumType::c_type sum = 0;
+
+ private:
+  template <int64_t kDenseRoundSize>
+  ThisType ConsumeDense(const T* values, const int64_t length) const {
```

Comment:
Why is it called "dense"? Perhaps "NoNulls" instead?

Reply:
Done